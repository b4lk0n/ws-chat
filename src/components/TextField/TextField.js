import React, { useCallback } from 'react';
import cx from 'classnames';
import { isFunction } from 'lodash-es';
import { string, bool } from 'prop-types';
import { Icon } from '../Icon';
import { IconButton } from '../IconButton';

import styles from './TextField.module.css';

export const TextField = ({ icon, withClear, onChange, className, ...props }) => {
  const handleChange = useCallback(
    e => {
      if (isFunction(onChange)) {
        onChange(e.target.value, e);
      }
    },
    [onChange]
  );
  const handleClear = useCallback(() => {
    if (isFunction(onChange)) {
      onChange('');
    }
  }, [onChange]);

  return (
    <div
      className={cx(
        styles.root,
        {
          [styles.withIcon]: !!icon,
          [styles.withClear]: withClear,
          [styles.disabled]: props.disabled,
        },
        className
      )}
    >
      <input {...props} onChange={handleChange} />

      {icon && <Icon value={icon} className={styles.icon} />}
      {withClear && <IconButton icon={'cancel'} size={'small'} onClick={handleClear} className={styles.clearButton} />}
    </div>
  );
};
TextField.propTypes = {
  icon: string,
  withSearch: bool,
  className: string,
};
