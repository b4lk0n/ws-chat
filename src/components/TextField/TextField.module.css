.root {
  display: inline-block;
  height: 40px;
  border: 1px solid var(--divider);
  border-radius: 4px;
  overflow: hidden;
  position: relative;
  transition: border-color 0.15s ease-in-out;
}

.root:focus-within {
  border-color: var(--textSecondary);
}

.root > input {
  padding: 10px 16px;
  border: none;
  outline: none;
  min-width: 200px;
  height: 100%;
  color: var(--textPrimary);
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen', 'Ubuntu', 'Cantarell', 'Fira Sans',
    'Droid Sans', 'Helvetica Neue', sans-serif;
  font-size: 14px;
  line-height: 20px;
  font-weight: 400;
  letter-spacing: 0.25px;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

.root.withIcon > input {
  padding-left: 40px;
}

.root.withClear > input {
  padding-right: 34px;
}

.root > input::placeholder,
.icon,
.clearButton {
  color: var(--textSecondary);
}

.root.disabled {
  pointer-events: none;
}

.root.disabled > input {
  color: var(--textSecondary);
}

.root.disabled > input::placeholder,
.root.disabled > .icon,
.root.disabled > .clearButton {
  color: var(--divider);
}

.icon,
.clearButton {
  position: absolute;
}

.icon {
  top: 8px;
  left: 8px;
}

.clearButton {
  top: 11px;
  right: 8px;
}
