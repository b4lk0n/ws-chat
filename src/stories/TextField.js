import React, { useState } from 'react';
import { withKnobs, text, boolean, number } from '@storybook/addon-knobs';
import { action } from '@storybook/addon-actions';
import { TextField } from '../components';

export default {
  title: 'TextField',
  decorators: [withKnobs],
};

export const Default = () => {
  const [value, setValue] = useState('');

  return (
    <TextField
      placeholder={text('placeholder', 'Enter any textâ€¦')}
      value={value}
      onChange={setValue}
      icon={text('icon', 'search')}
      withClear={boolean('withClear', false)}
      disabled={boolean('disabled', false)}
    />
  );
};
