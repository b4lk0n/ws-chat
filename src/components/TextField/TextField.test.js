import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { TextField } from './TextField';

describe('<TextField />', () => {
  it('should render only input by default', () => {
    const { container, getByPlaceholderText } = render(<TextField placeholder={'Text'} />);
    expect(getByPlaceholderText('Text')).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
    expect(container.firstChild.childNodes).toHaveLength(1);
    expect(container.firstChild.firstChild).toContainElement(container.firstChild.lastChild);
  });

  it('should add additional class if provided', () => {
    const { container } = render(<TextField className={'title'} />);
    expect(container.firstChild).toHaveClass('title');
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render icon', () => {
    const { container } = render(<TextField icon={'search'} />);
    expect(container.firstChild).toHaveClass('withIcon');
    expect(container.firstChild.childNodes).toHaveLength(2);
    expect(container.firstChild.lastChild).toHaveTextContent('search');
    expect(container.firstChild.lastChild).toHaveClass('material-icons', 'icon');
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should not render clear button for empty input', () => {
    const { container } = render(<TextField withClear />);
    expect(container.firstChild).toHaveClass('withClear');
    expect(container.firstChild.childNodes).toHaveLength(1);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render clear button for input with value', () => {
    const changeHandler = jest.fn();
    const { container } = render(<TextField withClear value={'Text'} onChange={changeHandler} />);
    expect(container.firstChild).toHaveClass('withClear');
    expect(container.firstChild.childNodes).toHaveLength(2);
    expect(container.firstChild.lastChild).toHaveClass('clearButton', 'small');
    fireEvent.click(container.firstChild.lastChild);
    expect(changeHandler).toBeCalledTimes(1);
    expect(changeHandler).toBeCalledWith('');
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render disabled elements', () => {
    const changeHandler = jest.fn();
    const { container } = render(<TextField withClear value={'Text'} onChange={changeHandler} disabled />);
    expect(container.firstChild).toHaveClass('disabled');
    fireEvent.click(container.firstChild.lastChild);
    expect(changeHandler).not.toBeCalled();
    expect(container.firstChild).toMatchSnapshot();
  });
});
