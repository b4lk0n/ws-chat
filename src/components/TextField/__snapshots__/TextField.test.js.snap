// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> should add additional class if provided 1`] = `
<div
  class="root title"
>
  <input />
</div>
`;

exports[`<TextField /> should not render clear button for empty input 1`] = `
<div
  class="root withClear"
>
  <input />
</div>
`;

exports[`<TextField /> should render clear button for input with value 1`] = `
<div
  class="root withClear"
>
  <input
    value="Text"
  />
  <button
    class="root small clearButton"
  >
    <span
      class="root small material-icons"
    >
      cancel
    </span>
  </button>
</div>
`;

exports[`<TextField /> should render disabled elements 1`] = `
<div
  class="root withClear disabled"
>
  <input
    disabled=""
    value="Text"
  />
  <button
    class="root small clearButton"
  >
    <span
      class="root small material-icons"
    >
      cancel
    </span>
  </button>
</div>
`;

exports[`<TextField /> should render icon 1`] = `
<div
  class="root withIcon"
>
  <input />
  <span
    class="root undefined material-icons icon"
  >
    search
  </span>
</div>
`;

exports[`<TextField /> should render only input by default 1`] = `
<div
  class="root"
>
  <input
    placeholder="Text"
  />
</div>
`;
